### PR TITLE
OOP school library: add Person, Student and Teacher classes

### PR DESCRIPTION
**Pull request by Kanyorok**
- Implement classes and objects in Ruby.
- Implement encapsulation and inheritance with Ruby.
- Run a program using the command line.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+# class for library
+class Person
+  # constructor method
+  def initialize(id, name = 'unknown', parent_permission: true)
```
Comment (Efiamotu-1) [inline person.rb]: - [ ] you missed providing the `age` parameter for the constructor function please kindly add this :+1: 
- [ ] Also i see that you are passing in the `id` as a parameter, this should be auto generated. see the code snippet below as a guide
- code snippet
```ruby
  def initialize(age, name = "Unknown", parent_permission = true)
    @id = Random.rand(1..1000)
    @age = age
    @name = name
    @parent_permission = parent_permission
  end
```

File: students.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+# subclass of the class person
+class Students < Person
+  def initialize(id, age, classroom, name: 'Unknown', parent_permission: true)
+    super(id, age, name: name, parent_permission: parent_permission)
```
Comment (Efiamotu-1) [inline students.rb]: - [ ] Kindly see the code snippet below for how to better structure your initialize method :+1: 
- code snippet
```ruby
def initialize(age, classroom, name = "Unknown", parent_permission = true)  
    super(age, name, parent_permission)
    @classroom = classroom
  end
```

File: teachers.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+# subclass for the person
+class Teacher < Person
+  def initialize(id, age, specialization, name: 'Unknown', parent_permission: true)
```
Comment (Efiamotu-1) [inline teachers.rb]: - [ ] Please kindly structure this as well with the example codes i have provided to you :+1:

**Review by Efiamotu-1 — CHANGES_REQUESTED** [review]
Hi @Kanyorok, 

 

Good job so far! 👏. Your Code is Awesome! 💪

There are some **few** issues 🐛 that you still need to work 🔧 on though to go to the next project ⏭️ but you are almost there! 🏃
## Don't Worry 🙏

It's ok to have bugs in your code 🐛. Please, see this code review as a **chance** for you to **learn** practical coding skills and learn **new tips** that can improve your code quality 💯 and help you **get a better ** 💼.

_I am going to guide you on how to fix them down below. 🤓_

## Highlights :dart: 
- ✔️ No linter Errors
- ✔️ Descriptive README
- ✔️ Descriptive PR title and description
- The code is top-notch! 💯 and has a huge potential.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

> **Note:** Please, don't Merge This Pull Request yet. Only Merge when it is Approved ✅.


## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [@Efiamotu-1](https://github.com/Efiamotu-1/)
- This is My **Twitter**: [@EFYAMOTU](https://twitter.com/EFYAMOTU)
- This is My **linkedIn**.[Habeeb Efiamotu Musa](https://www.linkedin.com/in/musa-habeeb/)
- This is My **Slack**: [Habeeb Efiamotu Musa](https://slack.com)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb (hunk 1 above)
Comment (Kanyorok) [reply person.rb]: @Efiamotu-1, thank you for the review. However, I do not understand this comment " You missed providing the name parameter for the constructor function please kindly add this 👍". Could you clarify the issue? As you can see from the screenshot you sent, the name parameter is already added in the constructor. `def initialize(id, name = 'unknown', parent_permission: true)`

Comment (Efiamotu-1) [reply person.rb]: sorry @Kanyorok i meant to type `age`

Comment (Kanyorok) [reply person.rb]: Ooh, thank you @Efiamotu-1. I am fixing the issue.

File: students.rb [hunk 4]
```diff
@@ -2,8 +2,10 @@
 
 # subclass of the class person
 class Students < Person
+  attr_reader :classroom
+
   def initialize(id, age, classroom, name: 'Unknown', parent_permission: true)
-    super(id, age, name: name, parent_permission: parent_permission)
+    super(id, age, name, parent_permission)
     @classroom = classroom
   end
```
Comment (abdulhamiid) [inline students.rb]: The student class Inherits from `Person`, where the `id` is auto-generated and not present in the constructor

So as requested by the previous reviewer, please remove the `id` from the parameters 😊

Your code 
```ruby
def initialize(id, age, classroom, name: 'Unknown', parent_permission: true)
super(id, age, name: name, parent_permission: parent_permission)
```

expected
```ruby
def initialize(age, classroom, name = "Unknown", parent_permission = true)
super(age, name, parent_permission)
```

🔗 Link to a lesson on [Inheritance](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#inheritance)

File: teachers.rb [hunk 5]
```diff
@@ -2,8 +2,10 @@
 
 # subclass for the person
 class Teacher < Person
+  attr_reader :specialization
+
   def initialize(id, age, specialization, name: 'Unknown', parent_permission: true)
-    super(id, age, name: name, parent_permission: parent_permission)
+    super(id, age, name, parent_permission)
     @specialization = specialization
   end
```
Comment (abdulhamiid) [inline teachers.rb]: As requested by the previous reviewer, please structure this as well by removing the `id` as the example codes provided to you above 👍

**Review by abdulhamiid — CHANGES_REQUESTED** [review]
# Status: Required Changes :recycle:
Hi @Kanyorok 👋, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, I really appreciate the effort you put in! but there are still some that haven't been addressed yet.

almost-there

## Highlights 🌟

- ✔️ Implement classes and objects in Ruby.
- ✔️ Implement encapsulation and inheritance with Ruby.

## Required Changes ♻️
_Please kindly check the comments under the review._

## Optional suggestions 🤷‍♂️
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

Cheers, and Happy coding!👏👏👏

## Important Notes 📝
- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Ibizugbe — APPROVED** [review]
Hi @Kanyorok 👋🏼👋🏼👋🏼, 

approved


Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights✅✅✅

- [x] The person and student class has been updated to match the requirements👍🏼.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**